File: LeetCode/33.Search_in_Rotated_Sorted_Array/33.Search_in_Rotated_Sorted_Array.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        int res = -1;

        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            }

            if (nums[mid] >= nums[left]) {
                if (target < nums[mid] && target >= nums[left]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (target > nums[mid] && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return res;
    }
}

// find pivot(most samllest point), then binary search two side
class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        int left = 0, right = n - 1;

        // find smallest element - pivot
        while (left <= right) {
            int m = (left + right) / 2;
            if (nums[m] > nums[n - 1]) {
                left = m + 1;
            } else {
                right = m - 1;
            }
        }

        // binary search left side, then right side
        int res = binarySearch(nums, 0, left - 1, target);
        if (res != -1)
            return res;

        return binarySearch(nums, left, n - 1, target);
    }

    private int binarySearch(int[] nums, int left, int right, int target) {
        while (left <= right) {
            int m = (left + right) / 2;
            if (nums[m] == target) {
                return m;
            } else if (nums[m] > target) {
                right = m - 1;
            } else {
                left = m + 1;
            }
        }
        return -1;
    }
}

File: LeetCode/33.Search_in_Rotated_Sorted_Array/33.Search_in_Rotated_Sorted_Array.swift
class Solution {
    func search(_ nums: [Int], _ target: Int) -> Int {
        let n = nums.count
        var l = 0, r = n
        // find pivot
        while l < r {
            let m = (l + r) / 2
            if nums[m] > nums[n - 1] {
                l = m + 1
            } else {
                r = m
            }
        }

        func binarySearch(_ l: Int, _ r: Int) -> Int {
            guard l <= r else { return -1 }

            var l = l, r = r
            while l < r {
                let m = (l + r) / 2
                if nums[m] == target {
                    return m
                } else if nums[m] < target {
                    l = m + 1
                } else {
                    r = m
                }
            }
            return -1
        }

        // [0..l-1] / [ l, n-1]
        if target > nums[n - 1] {
            return binarySearch(0, l)
        } else {
            return binarySearch(l, n)
        }

        return -1
    }
}
